openrtb_ext/request_wrapper.go

```diff
@@ -0,0 +1,757 @@
+	}
+
+	if err := rw.syncDeviceExt(); err != nil {
+		return err
+	}
+
+	if err := rw.syncRequestExt(); err != nil {
+		return err
+	}
+	if err := rw.syncAppExt(); err != nil {
+		return err
+	}
+	if err := rw.syncRegExt(); err != nil {
+		return err
+	}
+	if err := rw.syncSiteExt(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rw *RequestWrapper) syncUserExt() error {
+	if rw.Request.User == nil && rw.UserExt != nil && rw.UserExt.Dirty() {
+		rw.Request.User = &openrtb2.User{}
+	}
+	if rw.UserExt != nil && rw.UserExt.Dirty() {
+		userJson, err := rw.UserExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.User.Ext = userJson
+	}
+	return nil
+}
+
+func (rw *RequestWrapper) syncDeviceExt() error {
+	if rw.Request.Device == nil && rw.DeviceExt != nil && rw.DeviceExt.Dirty() {
+		rw.Request.Device = &openrtb2.Device{}
+	}
+	if rw.DeviceExt != nil && rw.DeviceExt.Dirty() {
+		deviceJson, err := rw.DeviceExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.Device.Ext = deviceJson
+	}
+	return nil
+}
+
+func (rw *RequestWrapper) syncRequestExt() error {
+	if rw.RequestExt != nil && rw.RequestExt.Dirty() {
+		requestJson, err := rw.RequestExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.Ext = requestJson
+	}
+	return nil
+}
+
+func (rw *RequestWrapper) syncAppExt() error {
+	if rw.Request.App == nil && rw.AppExt != nil && rw.AppExt.Dirty() {
+		rw.Request.App = &openrtb2.App{}
+	}
+	if rw.AppExt != nil && rw.AppExt.Dirty() {
+		appJson, err := rw.AppExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.App.Ext = appJson
+	}
+	return nil
+}
+
+func (rw *RequestWrapper) syncRegExt() error {
+	if rw.Request.Regs == nil && rw.RegExt != nil && rw.RegExt.Dirty() {
+		rw.Request.Regs = &openrtb2.Regs{}
+	}
+	if rw.RegExt != nil && rw.RegExt.Dirty() {
+		regsJson, err := rw.RegExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.Regs.Ext = regsJson
+	}
+	return nil
+}
+
+func (rw *RequestWrapper) syncSiteExt() error {
+	if rw.Request.Site == nil && rw.SiteExt != nil && rw.SiteExt.Dirty() {
+		rw.Request.Site = &openrtb2.Site{}
+	}
+	if rw.SiteExt != nil && rw.SiteExt.Dirty() {
+		siteJson, err := rw.SiteExt.Marshal()
+		if err != nil {
+			return err
+		}
+		rw.Request.Regs.Ext = siteJson
+	}
+	return nil
+}
+
+// ---------------------------------------------------------------
+// UserExt provides an interface for request.user.ext
+// ---------------------------------------------------------------
+
+type UserExt struct {
+	ext            map[string]json.RawMessage
+	extDirty       bool
+	consent        *string
+	consentDirty   bool
+	prebid         *ExtUserPrebid
+	prebidDirty    bool
+	digiTrust      *ExtUserDigiTrust
+	digiTrustDirty bool
+	eids           *[]ExtUserEid
+	eidsDirty      bool
+}
+
+func (ue *UserExt) Unmarshal(extJson json.RawMessage) error {
+	ue.ext = make(map[string]json.RawMessage)
+	ue.eids = &[]ExtUserEid{}
+	if len(extJson) == 0 || len(ue.ext) != 0 {
```

Comment:
No, if we submit an empty json, we want to generate an empty UserExt instance. It is a valid use case for a request to come in without a UserExt, but we still want to read that the elements are not there, and potentially write to them. Basically after getting rid of Extract(), we want Unmarshal() to ensure that we have a properly initialized UserExt even when the original request did not include json to be extracted.

Reply:
Ok that makes sense, however, shouldn't line 222 be after this if block? As written, `ue.ext` will always be 0 since we're always initializing `ue.ext` to an empty map on line 222.

Reply:
Good catch, I have that issue in all of them. Also I will add a dirty check, incase you try a second unmarshalling of an empty ext after some code has started setting values.